### ISSUE TITLE
[FEATURE] sparse embedding哪时支持concat dynamic embed_dim in dim 2

### ISSUE THREAD

**Issue by XFR1998** [datawhalechina/torch-rechub#226]
# TODO: support concat dynamic embed_dim in dim 2
            # [batch_size, num_features, embed_dim]
            sparse_emb = torch.cat(sparse_emb, dim=1)

**Comment by 1985312383** [comment]
这个部分不太好极简设计，暂时还没想好怎么做，可以保证动态emb_dim和所有特征兼容放在emb_dict里

**Comment by XFR1998** [comment]
class EmbeddingLayer(nn.Module)里并没有实现返回        - Sequence: same as sparse or ``(B, num_seq, L, embed_dim)`` when ``pooling="concat"``.
即使在pooling="concat"情况下，最终还是被sparse_emb = torch.cat(sparse_emb, dim=1)，有点不合理吧？

**Comment by zerolovesea** [comment]
> class EmbeddingLayer(nn.Module)里并没有实现返回 - Sequence: same as sparse or `(B, num_seq, L, embed_dim)` when `pooling="concat"`. 即使在pooling="concat"情况下，最终还是被sparse_emb = torch.cat(sparse_emb, dim=1)，有点不合理吧？

concat pooling做feature间的concat，这个感觉ok的吧

**Comment by XFR1998** [comment]
> > class EmbeddingLayer(nn.Module)里并没有实现返回 - Sequence: same as sparse or `(B, num_seq, L, embed_dim)` when `pooling="concat"`. 即使在pooling="concat"情况下，最终还是被sparse_emb = torch.cat(sparse_emb, dim=1)，有点不合理吧？
> 
> concat pooling做feature间的concat，这个感觉ok的吧

感谢回复！按现在的逻辑，如果pooling="concat"，那岂不是sparse特征(bs, 1, emb_dim)和sequence特征(bs, seq_len, emb_dim)直接在dim=1 concat了？相当于sequence特征的每个位置当成一个sparse特征和sparse特征混在一起了，这是合理的吗

**Comment by 1985312383** [comment]
> > > class EmbeddingLayer(nn.Module)里并没有实现返回 - Sequence: same as sparse or `(B, num_seq, L, embed_dim)` when `pooling="concat"`. 即使在pooling="concat"情况下，最终还是被sparse_emb = torch.cat(sparse_emb, dim=1)，有点不合理吧？
> > 
> > 
> > concat pooling做feature间的concat，这个感觉ok的吧
> 
> 感谢回复！按现在的逻辑，如果pooling="concat"，那岂不是sparse特征(bs, 1, emb_dim)和sequence特征(bs, seq_len, emb_dim)直接在dim=1 concat了？相当于sequence特征的每个位置当成一个sparse特征和sparse特征混在一起了，这是合理的吗

当前 pooling="concat" 并不是把 sequence feature 池化成一个普通 sparse-like feature，而是保留 [B, L, D] 的序列维度，外层再 unsqueeze(1) 后，单个序列特征会变成 [B, 1, L, D]。它主要给序列模型(DIN/DIEN/BST/MIND/GRU4Rec)单独调用，比如 embedding(x, self.history_features).squeeze(1)，而不是和普通 sparse feature 混在同一个 features 列表里做 torch.cat。

mean/sum pooling 后的 sequence 可以和 sparse 一起拼；concat pooling 应该走序列特征专用路径。如果混用 sparse [B,1,D] 和 concat sequence [B,1,L,D]，语义确实不合理，应该通过文档和代码校验限制掉。

后面我会试着增加判断校验来区分序列特征的处理方式，文档上明确 pooling="concat" 是序列模型专用路径

**Comment by zerolovesea** [comment]
> > > class EmbeddingLayer(nn.Module)里并没有实现返回 - Sequence: same as sparse or `(B, num_seq, L, embed_dim)` when `pooling="concat"`. 即使在pooling="concat"情况下，最终还是被sparse_emb = torch.cat(sparse_emb, dim=1)，有点不合理吧？
> > 
> > 
> > concat pooling做feature间的concat，这个感觉ok的吧
> 
> 感谢回复！按现在的逻辑，如果pooling="concat"，那岂不是sparse特征(bs, 1, emb_dim)和sequence特征(bs, seq_len, emb_dim)直接在dim=1 concat了？相当于sequence特征的每个位置当成一个sparse特征和sparse特征混在一起了，这是合理的吗

sequence特征也是要先做pooling 再concat的，直接concat不行

**Comment by XFR1998** [comment]
嗯如果是想序列特征pooling=concat和sparse一起用，那就只能分别过一次

        input_user = self.embedding(x, self.user_features, squeeze_dim=True)
        history_emb = self.embedding(x, self.history_features).squeeze(1)

**Comment by 1985312383** [comment]
> 嗯如果是想序列特征pooling=concat和sparse一起用，那就只能分别过一次
> 
> ```
>     input_user = self.embedding(x, self.user_features, squeeze_dim=True)
>     history_emb = self.embedding(x, self.history_features).squeeze(1)
> ```

是的
``` python
input_user = self.embedding(x, self.user_features, squeeze_dim=True)  # [B, D]
history_emb = self.embedding(x, self.history_features).squeeze(1)     # [B, L, D]
```
